lib/portage/dbapi/bintree.py

```diff
@@ -2183,22 +2166,73 @@ def _allocate_filename_multi(self, cpv):
             )
 
         if binpkg_format == "xpak":
-            filename_format = "%s-%s.xpak"
+            filename_format = "{}-{}.xpak"
         elif binpkg_format == "gpkg":
-            filename_format = "%s-%s.gpkg.tar"
+            filename_format = "{}-{}.gpkg.tar"
         else:
             raise InvalidBinaryPackageFormat(binpkg_format)
 
         while True:
-            filename = filename_format % (
+            filename = filename_format.format(
                 os.path.join(self.pkgdir, cpv.cp, pf),
                 build_id,
             )
             if os.path.exists(filename):
                 build_id += 1
             else:
+                try:
+                    # Avoid races
+                    ensure_dirs(os.path.dirname(filename))
+                    with open(filename, "x") as f:
+                        pass
+                except FileExistsError:
+                    build_id += 1
+                    continue
                 return filename
 
+    def _allocate_build_id(self, cpv):
```

Comment:
Actually, why are there two almost-identical functions? How about using one function that returns a tuple of `(filename, build_id)`?

Reply:
Put the 2 functions together, but create 2 more warp functions that return `filename` and `(filename, build_id)`. Since `getname`used in many place that will broke other tools.